contracts/incentives/base/BaseIncentivesController.sol

```diff
@@ -0,0 +1,207 @@
+  {
+    require(assets.length == emissionsPerSecond.length, 'INVALID_CONFIGURATION');
+
+    DistributionTypes.AssetConfigInput[] memory assetsConfig =
+      new DistributionTypes.AssetConfigInput[](assets.length);
+
+    for (uint256 i = 0; i < assets.length; i++) {
+      assetsConfig[i].underlyingAsset = assets[i];
+      assetsConfig[i].emissionPerSecond = uint104(emissionsPerSecond[i]);
+
+      require(assetsConfig[i].emissionPerSecond == emissionsPerSecond[i], 'INVALID_CONFIGURATION');
+
+      assetsConfig[i].totalStaked = IScaledBalanceToken(assets[i]).scaledTotalSupply();
+    }
+    _configureAssets(assetsConfig);
+  }
+
+  /// @inheritdoc IAaveIncentivesController
+  function handleAction(
+    address user,
+    uint256 totalSupply,
+    uint256 userBalance
+  ) external override {
+    uint256 accruedRewards = _updateUserAssetInternal(user, msg.sender, userBalance, totalSupply);
+    if (accruedRewards != 0) {
+      _usersUnclaimedRewards[user] = _usersUnclaimedRewards[user].add(accruedRewards);
+      emit RewardsAccrued(user, accruedRewards);
+    }
+  }
+
+  /// @inheritdoc IAaveIncentivesController
+  function getRewardsBalance(address[] calldata assets, address user)
+    external
+    view
+    override
+    returns (uint256)
+  {
+    uint256 unclaimedRewards = _usersUnclaimedRewards[user];
+
+    DistributionTypes.UserStakeInput[] memory userState =
+      new DistributionTypes.UserStakeInput[](assets.length);
+    for (uint256 i = 0; i < assets.length; i++) {
+      userState[i].underlyingAsset = assets[i];
+      (userState[i].stakedByUser, userState[i].totalStaked) = IScaledBalanceToken(assets[i])
+        .getScaledUserBalanceAndSupply(user);
+    }
+    unclaimedRewards = unclaimedRewards.add(_getUnclaimedRewards(user, userState));
+    return unclaimedRewards;
+  }
+
+  /// @inheritdoc IAaveIncentivesController
+  function claimRewards(
+    address[] calldata assets,
+    uint256 amount,
+    address to
+  ) external override returns (uint256) {
+    require(to != address(0), 'INVALID_TO_ADDRESS');
+    return _claimRewards(assets, amount, msg.sender, msg.sender, to);
+  }
+
+  /// @inheritdoc IAaveIncentivesController
+  function claimRewardsOnBehalf(
+    address[] calldata assets,
+    uint256 amount,
+    address user,
+    address to
+  ) external override onlyAuthorizedClaimers(msg.sender, user) returns (uint256) {
+    require(user != address(0), 'INVALID_USER_ADDRESS');
+    require(to != address(0), 'INVALID_TO_ADDRESS');
+    return _claimRewards(assets, amount, msg.sender, user, to);
+  }
+
+  /// @inheritdoc IAaveIncentivesController
+  function claimRewardsToSelf(address[] calldata assets, uint256 amount)
+    external
+    override
+    returns (uint256)
+  {
+    return _claimRewards(assets, amount, msg.sender, msg.sender, msg.sender);
+  }
+
+  /// @inheritdoc IAaveIncentivesController
+  function setClaimer(address user, address caller) external override onlyEmissionManager {
```

Comment:
Should it be possible for a user to set a claimer for themselves? e.g., update `_authorizedClaimers[msg.sender]`, don't see why it needs to be the EmissionManager only, as long as it cannot be set on behalf of other users.

Reply:
I think this function makes sense as it is since its only callable by the EmissionManager.

Reply:
This function `function setClaimer(address user, address caller)` is intented for 3rd party smart contracts that doesnt support claim from scratch and needs another contract as the incentives distributor for their users. Thats why EmissionManager trusted role is used here.

Would be interesting that end users could also set their own claimers, like a external hot wallet that claims rewards to the cold wallet address, but i would not like to change the current Abstract incentives contract in this PR. Could be added in a separate PR as this one is quite big and only focused into making incentives logic abstract.

Reply:
Can you create an issue to track your optimization? Thanks

Reply:
Added in #9